archiver/src/repack.rs

```diff
@@ -0,0 +1,80 @@
+use crate::{error::Result, ArchiveKind, Archiver, Unarchiver};
+use std::{io, path::PathBuf};
+use tempfile::TempDir;
+
+/// Very naive implementation of a repacker
+pub struct Repacker {
+    unarchiver: Unarchiver,
+    _work: TempDir, // FIXME: avoid temporary directory
```

Comment:
I think you could probably add one layer of abstraction here.

```rust
struct DirEntry<R> {
  path: PathBuf,
  source: R
}
fn read_dir(path: impl AsRef<Path>) -> io::Result<impl Iterator<Item = DirEntry<impl Read>>;
fn write_dir(entries: impl Iterator<Item = DirEntry<impl Read>>) -> io::Result<()>;
```

and instead of having the Archiver take in a pathbuf, have it take in a `Iterator<Item = DirEntry<R>>` and have the Unarchiver spit out an `Iterator<Item = DirEntry<R>>` and just have them pipe it into wherever. If you buffer tbe entries passed into `write_dir`, you might even be able to easily parallelize that bit

Reply:
I really like your idea. Let's implement that in a later PR ?